File: Algorithms/Graphs/Python/Dijkstra.py
class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = [[0] * vertices for _ in range(vertices)]

    def print_solution(self, dist):
        print("Vertex \t Distance from Source")
        for node in range(self.V):
            print(node, "\t\t", dist[node])

    def min_distance(self, dist, visited):
        min_dist = float("inf")
        min_dist_index = None
        for v in range(self.V):
            if dist[v] < min_dist and not visited[v]:
                min_dist = dist[v]
                min_dist_index = v
        return min_dist_index

    def dijkstra(self, source):
        dist = [float("inf")] * self.V
        dist[source] = 0
        visited = [False] * self.V

        for _ in range(self.V):
            u = self.min_distance(dist, visited)
            visited[u] = True
            for v in range(self.V):
                if (
                    self.graph[u][v] > 0
                    and not visited[v]
                    and dist[v] > dist[u] + self.graph[u][v]
                ):
                    dist[v] = dist[u] + self.graph[u][v]

        self.print_solution(dist)


# Example usage:
# Create a graph with 6 vertices
g = Graph(6)

# Set up the graph connections and weights
g.graph = [
    [0, 1, 4, 0, 0, 0],
    [1, 0, 2, 7, 0, 0],
    [4, 2, 0, 1, 5, 0],
    [0, 7, 1, 0, 3, 2],
    [0, 0, 5, 3, 0, 6],
    [0, 0, 0, 2, 6, 0],
]

# Perform Dijkstra's algorithm starting from vertex 0
g.dijkstra(0)

File: Algorithms/Graphs/C/Dijkstra.c
#include <stdio.h>
#include <stdbool.h>
#include <limits.h>

#define V 6 // Number of vertices

int minDistance(int dist[], bool visited[])
{
    int min = INT_MAX;
    int min_index = -1;

    for (int v = 0; v < V; v++)
    {
        if (!visited[v] && dist[v] < min)
        {
            min = dist[v];
            min_index = v;
        }
    }

    return min_index;
}

void printSolution(int dist[])
{
    printf("Vertex\tDistance from Source\n");
    for (int i = 0; i < V; i++)
    {
        printf("%d\t%d\n", i, dist[i]);
    }
}

void dijkstra(int graph[V][V], int source)
{
    int dist[V];
    bool visited[V];

    for (int i = 0; i < V; i++)
    {
        dist[i] = INT_MAX;
        visited[i] = false;
    }

    dist[source] = 0;

    for (int count = 0; count < V - 1; count++)
    {
        int u = minDistance(dist, visited);
        visited[u] = true;

        for (int v = 0; v < V; v++)
        {
            if (!visited[v] && graph[u][v] && dist[u] != INT_MAX && dist[u] + graph[u][v] < dist[v])
            {
                dist[v] = dist[u] + graph[u][v];
            }
        }
    }

    printSolution(dist);
}

int main()
{
    int graph[V][V] = {
        {0, 1, 4, 0, 0, 0},
        {1, 0, 2, 7, 0, 0},
        {4, 2, 0, 1, 5, 0},
        {0, 7, 1, 0, 3, 2},
        {0, 0, 5, 3, 0, 6},
        {0, 0, 0, 2, 6, 0}};

    int source = 0;

    dijkstra(graph, source);

    return 0;
}
